src/MSBuild/TerminalLogger/TerminalLogger.cs

```diff
@@ -635,28 +635,38 @@ private bool IsImmediateMessage(string message)
     /// </summary>
     private void ErrorRaised(object sender, BuildErrorEventArgs e)
     {
-        var buildEventContext = e.BuildEventContext;
-        if (buildEventContext is not null && _projects.TryGetValue(new ProjectContext(buildEventContext), out Project? project))
-        {
-            string message = EventArgsFormatting.FormatEventMessage(
+        BuildEventContext? buildEventContext = e.BuildEventContext;
+        Project? project = null;
+        bool isTrackedProject = buildEventContext is not null && _projects.TryGetValue(new ProjectContext(buildEventContext), out project);
+        string message = EventArgsFormatting.FormatEventMessage(
                 category: AnsiCodes.Colorize("error", TerminalColor.Red),
                 subcategory: e.Subcategory,
                 message: e.Message,
                 code: AnsiCodes.Colorize(e.Code, TerminalColor.Red),
                 file: HighlightFileName(e.File),
-                projectFile: null,
+
+                // for the tracked projects the project file name is included in the final output result.
+                projectFile: isTrackedProject ? null : e.ProjectFile ?? string.Empty,
                 lineNumber: e.LineNumber,
                 endLineNumber: e.EndLineNumber,
                 columnNumber: e.ColumnNumber,
                 endColumnNumber: e.EndColumnNumber,
                 threadId: e.ThreadId,
                 logOutputProperties: null);
 
-            project.AddBuildMessage(MessageSeverity.Error, message);
+        if (isTrackedProject)
+        {
+            project!.AddBuildMessage(MessageSeverity.Error, message);
```

Comment:
This statement makes us print something like `failed with errors` at the end of the build (end of restore). If we print the error as an immediate message, we lose this information and the final output doesn't mention errors. Is this acceptable?

Reply:
the failed message is presented , because it happens during restore phase 
![image](https://github.com/dotnet/msbuild/assets/95473390/cce2a172-ad94-4025-a1b4-81a563c8f472)

Reply:
Correct, I am specifically interested in `failed` vs. `failed with errors`, as determined here: https://github.com/dotnet/msbuild/blob/85d842283030ac4f55511fe614fa5429f406c9d9/src/MSBuild/TerminalLogger/TerminalLogger.cs#L873-L878

If the build prints an error, I think the intended output would be `Restore failed with errors`.

Reply:
Looks like you could set `_buildHasErrors` in this block to fix this (and I think that's worth doing).